client/modules/cm-lang-ftml/src/grammars/ftml.yaml

```diff
@@ -0,0 +1,999 @@
+
+  ColoredText:
+    begin:
+      match: /(##)(#?[^\n\|]+)(\|)/
+      captures:
+        0: { open: ColoredText, tag: processingInstruction }
+        1: { type: ColoredTextColor, tag: color }
+        2: { type: ColoredTextSeparator, tag: separator }
+    end:
+      close: ColoredText
+      match: "##"
+      tag: processingInstruction
+
+  Typography:
+    tag: processingInstruction
+    match:
+      # ``quotation''
+      - /``(?=(?!``).+?'')/
+      - /(?<=``(?!'').+?)''/
+      # `quotation'
+      - /`(?=(?!`).+?')/
+      - /(?<=`(?!').+?)'/
+      # ,,quotation'' (this one is so damn stupid)
+      - /,,(?=(?!,,).+?'')/
+      - /(?<=,,(?!,,).+?)''/
+      # <<, >>
+      - /<<|>>/
+      # ...
+      - /\.{3}/
+      # --
+      - /(?<=\s)--(?=\s)/
+
+  # -- BLOCK MARKUP
+
+  # symbols that interrupt a paragraph on line start
+  interrupt: /(?:\++\*?@ws)|(?:[\[\]])|(?:[*#]@ws)|(?:[-=]{4,})|(?::)|(?:>+@ws)|(?:>$)|(?:\|{2})/
+
+  hr: /(?:-{3,}|={3,})@ws*$/
+  heading: /(?:\++\*?)@ws+(?!$)/
+  cs: /@sws(?:(?:>+@ws|>|[*#]@ws)@ws*)+/
+
+  HorizontalRule:
+    match: /@sws@hr/
+    tag: contentSeparator
+
+  ClearFloat:
+    match: /@sws~{3,}[<>]?/
+    tag: keyword
+
+  Heading:
+    begin: { type: HeadingMark, match: /@sws@heading/, tag: heading }
+    end: { match: /$/ }
+    inside:
+      - include: inline
+      - include: blocks
+
+  Center:
+    begin: { type: CenterMark, match: /@sws=@ws+/, tag: heading }
+    end: { match: /$/ }
+    inside:
+      - include: inline
+      - include: blocks
+
+  Table:
+    fold: past_first_line
+    begin:
+      match: /@sws\|{2}/
+      rematch: true
+    end:
+      lookbehind: '!/@esc\s*/'
+      match: /@enl|^(?:(?!\|{2}).)+$/
+      rematch: true
+    inside:
+      - type: TableMark
+        match: /(\|{2,})([<=~>]?)/
+        captures:
+          0: { type: TableMarkSeparator, tag: separator }
+          1: { type: TableMarkModifier, tag: modifier }
+      - include: inline
+      - include: blocks
+
+  ContainerMark:
+    match: /@sws>+|[*#]/
+    tag: keyword
+
+  Container:
+    fold: past_first_line
+    begin:
+      match: /@cs/
+      rematch: true
+    end:
+      lookbehind: '!/@esc\s*/'
+      match: /@enl|(?!@cs)@sws/
+      rematch: true
+    inside:
+      # simplified block markup inside containers
+      # horizontal rule
+      - match: /(@cs)(@hr)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: HorizontalRule }
+      # clearfloat
+      - match: /(@cs)(~{3,}[<>]?)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: ClearFloat }
+      # heading
+      - match: /(@cs)(@heading)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: HeadingMark }
+      # center
+      - match: /(@cs)(=@ws+)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: CenterMark }
+      # everything else
+      - ContainerMark
+      - include: inline
+      - include: blocks
+
```

Comment:
For the recodd, what's going wrong with paragraphs? (I know they're a pain, I'm just curious)

Reply:
I uh... don't know lmao

I disabled this rule because while it _usually_ works, it occasionally does weird screwy matches on stuff it shouldn't. I don't know why exactly this happens. I'll probably get back to this at some point and see if I can make a better rule for this.

Reply:
Okay, just add a `TODO` then I guess.